src/Microsoft.Health.Fhir.Shared.Api/Controllers/ImportController.cs

```
@@ -79,9 +93,80 @@ public class ImportController : Controller
+                {
+                    var line = await reader.ReadLineAsync();
+                    while (line != null)
+                    {
+                        var importResource = importParser.Parse(index, 0, 0, line, ImportMode.IncrementalLoad);
+                        var key = importResource.ResourceWrapper.ToResourceKey(true);
+                        if (!keys.Add(key))
+                        {
+                            throw new RequestNotValidException(string.Format(Resources.ResourcesMustBeUnique, key));
+                        }
+
+                        resources.Add(importResource);
+                        line = await reader.ReadLineAsync();
+                        index++;
+                    }
+                }
+                else
+                {
+                    var str = await reader.ReadToEndAsync();
+                    var resource = await fhirParser.ParseAsync(str);
+                    var bundle = resource.ToResourceElement().ToPoco<Bundle>();
+                    foreach (var entry in bundle.Entry)
+                    {
+                        if (entry.Request?.Method != Bundle.HTTPVerb.PUT)
+                        {
+                            throw new RequestNotValidException(string.Format(Resources.InvalidBundleEntry, entry.Request?.Method, KnownRoutes.ImportBundle));
+                        }
+
+                        var importResource = importParser.Parse(index, 0, 0, entry.Resource, ImportMode.IncrementalLoad);
+                        var key = importResource.ResourceWrapper.ToResourceKey(true);
+                        if (!keys.Add(key))
+                        {
+                            throw new RequestNotValidException(string.Format(Resources.ResourcesMustBeUnique, key));
+                        }
+
+                        resources.Add(importResource);
+                        index++;
+                    }
+                }
+            }
+            catch (Exception)
```

This exception is too broad. 
If any type of exception is thrown, then it'll return a "Unable to parse resource at index=", given that it's not necessarily and parsing issue.

---

Per example, when a RequestNotValidException is thrown with a ResourcesMustBeUnique or an InvalidBundleEntry, the details of those exceptions are lost.

---

What exception type do you suggest?

---

I changed the exception. Please keep in mind that current logic is "fail fast", it exits on first found problem. This is intentional.